runtime/src/main/java/com/vaadin/quarkus/context/RouteScopedContext.java

```diff
@@ -0,0 +1,345 @@
+/*
+ * Copyright 2000-2018 Vaadin Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.vaadin.quarkus.context;
+
+import javax.enterprise.context.spi.Contextual;
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.spi.Bean;
+import javax.enterprise.inject.spi.BeanManager;
+
+import java.io.Serializable;
+import java.lang.annotation.Annotation;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.Unremovable;
+
+import com.vaadin.flow.component.ComponentUtil;
+import com.vaadin.flow.component.UI;
+import com.vaadin.flow.component.page.ExtendedClientDetails;
+import com.vaadin.flow.router.AfterNavigationEvent;
+import com.vaadin.flow.router.BeforeEnterEvent;
+import com.vaadin.flow.router.RouterLayout;
+import com.vaadin.flow.server.VaadinSession;
+import com.vaadin.quarkus.annotation.NormalRouteScoped;
+import com.vaadin.quarkus.annotation.NormalUIScoped;
+import com.vaadin.quarkus.annotation.RouteScopeOwner;
+import com.vaadin.quarkus.annotation.VaadinSessionScoped;
+
+import static javax.enterprise.event.Reception.IF_EXISTS;
+
+/**
+ * Context for {@link NormalRouteScoped NormalRouteScoped} beans.
+ */
+public class RouteScopedContext extends AbstractContext {
+
+    public abstract static class ContextualStorageManager
+            extends AbstractContextualStorageManager<RouteStorageKey> {
```

Comment:
There is a reason why it's nested.
It has private methods which should not be exposed anyhow.
Being a nested class of the RouteScopedContext it still allows to call them. These methods have to be package local at least for the class in a new file.

Also there is no reason to extract this class outside since:
-  this is an implementation detail for `RouteScopedContext`
- no-one else is using this class
- the readability is not improved when the class is extracted in the new file: the border of the class are obvious regardless whether it's a nested class or a standalone class. Its functionality is clearly localized inside the nested class. The readability is an issue when the class functionality is a lambda/anonymous class or directly inlined in the encloising class. This is not the case for nested classes.
- the same code with minor changes is inside CDI impl of `RouteScopedContext`. So it's consistent.

Reply:
Then why don't you have the implementation in the implementation instead of having an abstract class implementation of an abstract class that is not abstract as it doesn't need any more implementation.

Reply:
Good question.
Because it's a one of the number of hacks which I made to be able to test this.

In the perfect world the `abstract` class is not needed and `RouteContextualStorageManager` should be just `ContextualStorageManager` .
And it was.
But in the real world it's not possible to test this anyhow because this class should be `@VaadinSessionScoped` and quarkus is not aware about `@VaadinSessionScoped`  in the tests.
And I cannot just override `ContextualStorageManager` in the tests since there will be another problem: I already tried that.

This is the only way to mock `ContextualStorageManager` which is in fact should not be `abstract` but used directly.
Check the tests and you will see how hard was my journey to get some `ContextualStorageManager` instance working there.